### PR TITLE
fix(complete): match start bufname

### PR DESCRIPTION
**Pull request by glepnir**
filter bufname by using match text when complete add `f` flag.

### CODE REVIEW

**Review by zeertzjq — DISMISSED** [review]
This behavior is very strange. Filtering is based on full buffer name, but the actual matches added are parts of the buffer name, separate by `/` or `.`.

**Comment by glepnir** [comment]
the path names are somewhat unimportant. maybe it would make sense to just use tail to match?

File: src/nvim/insexpand.c [hunk 1]
```diff
@@ -3334,7 +3334,9 @@ static bool get_next_completion_match(int type, ins_compl_next_state_T *st, pos_
 static void get_next_bufname_token(void)
 {
   FOR_ALL_BUFFERS(b) {
-    if (b->b_p_bl && b->b_sfname != NULL) {
+    if (b->b_p_bl && b->b_sfname != NULL
+        && strncmp(b->b_sfname, compl_first_match->cp_str,
+                   strlen(compl_first_match->cp_str)) == 0) {
```
Comment (justinmk) [inline src/nvim/insexpand.c:3339]: strequal

File: test/functional/editor/completion_spec.lua [hunk 2]
```diff
@@ -1251,5 +1251,21 @@ describe('completion', function()
       {0:~                   }|
       {3:-- }{4:match 1 of 5}     |
     ]]}
+
+    feed('<Esc>')
+    command('bd '..bufname ..' '..hidden ..' | edit foo | edit fooA | edit Bar')
+    feed('ccf<C-n>')
+    screen:snapshot_util()
```
Comment (justinmk) [inline test/functional/editor/completion_spec.lua:1258]: remove

File: src/nvim/insexpand.c (hunk 1 above)
Comment (zeertzjq) [reply src/nvim/insexpand.c]: That will compare too many chars.

File: src/nvim/insexpand.c [hunk 3]
```diff
@@ -3335,24 +3335,27 @@ static void get_next_bufname_token(void)
...
-        if (dot && *p == NUL && *current != '.') {
-          len = (int)(dot - current);
-          p = dot + 1;
-        }
-        ins_compl_add(current, len, NULL, NULL, false, NULL, 0,
-                      p_ic ? CP_ICASE : 0, false);
-        if (*p == NUL) {
-          break;
+      char *tail = path_tail(b->b_sfname);
+      if (strncmp(tail, compl_first_match->cp_str, strlen(compl_first_match->cp_str)) == 0) {
```
Comment (zeertzjq) [inline src/nvim/insexpand.c:3339]: Still not entirely right (doesn't respect 'ignorecase'), but `compl_orig_text` should be better here.
```suggestion
      if (strncmp(tail, compl_orig_text, strlen(compl_orig_text)) == 0) {
```

Comment (glepnir) [reply src/nvim/insexpand.c]: oh thx for tip it's better

File: src/nvim/insexpand.c [hunk 4]
```diff
@@ -3335,24 +3335,27 @@ static void get_next_bufname_token(void)
...
-        if (dot && *p == NUL && *current != '.') {
-          len = (int)(dot - current);
-          p = dot + 1;
-        }
-        ins_compl_add(current, len, NULL, NULL, false, NULL, 0,
-                      p_ic ? CP_ICASE : 0, false);
-        if (*p == NUL) {
-          break;
+      char *tail = path_tail(b->b_sfname);
+      if (strncmp(tail, compl_orig_text, strlen(compl_orig_text)) == 0) {
```
[the remark this answers is not in the document]
Comment (zeertzjq) [reply src/nvim/insexpand.c:3339]: This will not work.

**Review by zeertzjq — COMMENTED** [review]
The behavior is still strange: filtering is based on path tail, but every component in a file name is added to matches.

**Comment by glepnir** [comment]
Does the file path names make sense? I feel like just using the tail is fine.

**Comment by zeertzjq** [comment]
To be consistent, only path tails should be added to matches.

**Comment by glepnir** [comment]
agree . do it later

**Pull request merged**